File: alcura_ipd_ext/alcura_ipd_extensions/report/ipd_consultation_notes/ipd_consultation_notes.py
"""IPD Consultation Notes — Script Report.

Lists all Patient Encounters linked to Inpatient Records via
``custom_linked_inpatient_record``, with bed/ward context from the IR.
Supports filtering by company, patient, practitioner, department, note
type, ward, and date range.
"""

from __future__ import annotations

import frappe
from frappe import _
from frappe.utils import getdate


def execute(filters: dict | None = None) -> tuple[list[dict], list[dict]]:
	filters = filters or {}
	columns = _get_columns()
	data = _get_data(filters)
	return columns, data


def _get_columns() -> list[dict]:
	return [
		{
			"fieldname": "encounter",
			"label": _("Encounter"),
			"fieldtype": "Link",
			"options": "Patient Encounter",
			"width": 140,
		},
		{
			"fieldname": "encounter_date",
			"label": _("Date"),
			"fieldtype": "Date",
			"width": 100,
		},
		{
			"fieldname": "patient",
			"label": _("Patient"),
			"fieldtype": "Link",
			"options": "Patient",
			"width": 120,
		},
		{
			"fieldname": "patient_name",
			"label": _("Patient Name"),
			"fieldtype": "Data",
			"width": 160,
		},
		{
			"fieldname": "practitioner_name",
			"label": _("Practitioner"),
			"fieldtype": "Data",
			"width": 150,
		},
		{
			"fieldname": "note_type",
			"label": _("Note Type"),
			"fieldtype": "Data",
			"width": 130,
		},
		{
			"fieldname": "chief_complaint",
			"label": _("Chief Complaint"),
			"fieldtype": "Data",
			"width": 200,
		},
		{
			"fieldname": "provisional_diagnosis",
			"label": _("Provisional Diagnosis"),
			"fieldtype": "Data",
			"width": 200,
		},
		{
			"fieldname": "inpatient_record",
			"label": _("Inpatient Record"),
			"fieldtype": "Link",
			"options": "Inpatient Record",
			"width": 140,
		},
		{
			"fieldname": "ward",
			"label": _("Ward"),
			"fieldtype": "Data",
			"width": 100,
		},
		{
			"fieldname": "room",
			"label": _("Room"),
			"fieldtype": "Data",
			"width": 100,
		},
		{
			"fieldname": "bed",
			"label": _("Bed"),
			"fieldtype": "Data",
			"width": 100,
		},
		{
			"fieldname": "docstatus",
			"label": _("Status"),
			"fieldtype": "Data",
			"width": 80,
		},
	]


def _get_data(filters: dict) -> list[dict]:
	pe = frappe.qb.DocType("Patient Encounter")
	ir = frappe.qb.DocType("Inpatient Record")

	query = (
		frappe.qb.from_(pe)
		.left_join(ir)
		.on(pe.custom_linked_inpatient_record == ir.name)
		.select(
			pe.name.as_("encounter"),
			pe.encounter_date,
			pe.patient,
			pe.patient_name,
			pe.practitioner,
			pe.practitioner_name,
			pe.custom_ipd_note_type.as_("note_type"),
			pe.custom_chief_complaint_text.as_("chief_complaint"),
			pe.custom_provisional_diagnosis_text.as_("provisional_diagnosis"),
			pe.custom_linked_inpatient_record.as_("inpatient_record"),
			ir.custom_current_ward.as_("ward"),
			ir.custom_current_room.as_("room"),
			ir.custom_current_bed.as_("bed"),
			pe.docstatus,
		)
		.where(pe.custom_ipd_note_type.isnotnull())
		.where(pe.custom_ipd_note_type != "")
		.where(pe.docstatus != 2)
		.orderby(pe.encounter_date, order=frappe.qb.desc)
		.orderby(pe.creation, order=frappe.qb.desc)
	)

	if filters.get("company"):
		query = query.where(pe.company == filters["company"])
	if filters.get("patient"):
		query = query.where(pe.patient == filters["patient"])
	if filters.get("practitioner"):
		query = query.where(pe.practitioner == filters["practitioner"])
	if filters.get("medical_department"):
		query = query.where(pe.medical_department == filters["medical_department"])
	if filters.get("note_type"):
		query = query.where(pe.custom_ipd_note_type == filters["note_type"])
	if filters.get("ward"):
		query = query.where(ir.custom_current_ward == filters["ward"])
	if filters.get("from_date"):
		query = query.where(pe.encounter_date >= getdate(filters["from_date"]))
	if filters.get("to_date"):
		query = query.where(pe.encounter_date <= getdate(filters["to_date"]))

	rows = query.run(as_dict=True)

	status_map = {0: "Draft", 1: "Submitted"}
	for row in rows:
		row["docstatus"] = status_map.get(row.get("docstatus"), "")

	return rows

File: alcura_ipd_ext/alcura_ipd_extensions/report/ipd_consultation_notes/ipd_consultation_notes.js
// Copyright (c) 2026, Alcura and contributors
// For license information, please see license.txt
//
// US-E3: IPD Consultation Notes report filters

frappe.query_reports["IPD Consultation Notes"] = {
	filters: [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
		},
		{
			fieldname: "patient",
			label: __("Patient"),
			fieldtype: "Link",
			options: "Patient",
		},
		{
			fieldname: "practitioner",
			label: __("Practitioner"),
			fieldtype: "Link",
			options: "Healthcare Practitioner",
		},
		{
			fieldname: "medical_department",
			label: __("Department"),
			fieldtype: "Link",
			options: "Medical Department",
		},
		{
			fieldname: "note_type",
			label: __("Note Type"),
			fieldtype: "Select",
			options: "\nAdmission Note\nProgress Note\nProcedure Note\nConsultation Note\nDischarge Summary",
		},
		{
			fieldname: "ward",
			label: __("Ward"),
			fieldtype: "Link",
			options: "Hospital Ward",
		},
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
		},
	],

	formatter(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		if (column.fieldname === "note_type" && value) {
			const plain = $("<div>").html(value).text().trim();
			const color_map = {
				"Admission Note": "blue",
				"Progress Note": "green",
				"Procedure Note": "orange",
				"Consultation Note": "purple",
				"Discharge Summary": "grey",
			};
			const color = color_map[plain] || "grey";
			value = `<span class="indicator-pill ${color}">${plain}</span>`;
		}

		if (column.fieldname === "docstatus") {
			const plain = $("<div>").html(value).text().trim();
			if (plain === "Draft") {
				value = '<span class="indicator-pill orange">Draft</span>';
			} else if (plain === "Submitted") {
				value = '<span class="indicator-pill blue">Submitted</span>';
			}
		}

		return value;
	},
};
